test_util/src/pty.rs

```diff
@@ -0,0 +1,437 @@
+          process_handle,
+          self.inner,
+          process_handle,
+          &mut duplicate_handle,
+          0,
+          0,
+          DUPLICATE_SAME_ACCESS,
+        ));
+
+        WinHandle::new(duplicate_handle)
+      }
+    }
+
+    pub fn as_raw(&self) -> HANDLE {
+      self.inner
+    }
+
+    pub fn take(mut self) -> HANDLE {
+      self.close_on_drop = false;
+      self.inner
+    }
+  }
+
+  unsafe impl Send for WinHandle {}
+  unsafe impl Sync for WinHandle {}
+
+  impl Drop for WinHandle {
+    fn drop(&mut self) {
+      if self.close_on_drop {
+        unsafe {
+          if !self.inner.is_null() && self.inner != INVALID_HANDLE_VALUE {
+            winapi::um::handleapi::CloseHandle(self.inner);
+          }
+        }
+      }
+    }
+  }
+
+  struct ProcThreadAttributeList {
+    buffer: Vec<u8>,
+  }
+
+  impl ProcThreadAttributeList {
+    pub fn new(console_handle: HPCON) -> Self {
+      unsafe {
+        // discover size required for the list
+        let mut size = 0;
+        let attribute_count = 1;
+        assert_eq!(
+          InitializeProcThreadAttributeList(
+            ptr::null_mut(),
+            attribute_count,
+            0,
+            &mut size
+          ),
+          FALSE
+        );
+
+        let mut buffer = vec![0u8; size];
+        let attribute_list_ptr = buffer.as_mut_ptr() as _;
+
+        assert_win_success!(InitializeProcThreadAttributeList(
+          attribute_list_ptr,
+          attribute_count,
+          0,
+          &mut size,
+        ));
+
+        const PROC_THREAD_ATTRIBUTE_PSEUDOCONSOLE: usize = 0x00020016;
+        assert_win_success!(UpdateProcThreadAttribute(
+          attribute_list_ptr,
+          0,
+          PROC_THREAD_ATTRIBUTE_PSEUDOCONSOLE,
+          console_handle,
+          std::mem::size_of::<HPCON>(),
+          ptr::null_mut(),
+          ptr::null_mut(),
+        ));
+
+        ProcThreadAttributeList { buffer }
+      }
+    }
+
+    pub fn as_mut_ptr(&mut self) -> LPPROC_THREAD_ATTRIBUTE_LIST {
+      self.buffer.as_mut_slice().as_mut_ptr() as *mut _
+    }
+  }
+
+  impl Drop for ProcThreadAttributeList {
+    fn drop(&mut self) {
+      unsafe { DeleteProcThreadAttributeList(self.as_mut_ptr()) };
+    }
+  }
+
+  fn create_pipe() -> (WinHandle, WinHandle) {
+    unsafe {
+      let mut read_handle = std::ptr::null_mut();
+      let mut write_handle = std::ptr::null_mut();
+
+      assert_win_success!(CreatePipe(
+        &mut read_handle,
+        &mut write_handle,
+        ptr::null_mut(),
+        0
+      ));
+
+      (WinHandle::new(read_handle), WinHandle::new(write_handle))
+    }
+  }
+
+  fn to_windows_str(str: &str) -> Vec<u16> {
+    use std::os::windows::prelude::OsStrExt;
+    std::ffi::OsStr::new(str)
+      .encode_wide()
+      .chain(Some(0))
+      .collect()
+  }
+
+  fn get_env_vars(env_vars: HashMap<String, String>) -> Vec<u16> {
+    // each environment variable is in the form `name=value\0`
```

Comment:
I believe there is another requirement, which is that the env vars are sorted alphabetically (by key).

Reply:
cf. https://docs.microsoft.com/en-us/windows/win32/procthread/changing-environment-variables

Reply:
They really do work hard to make things fun. Thanks! I will do this.